language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -146,17 +161,83 @@ instance PrettyBy config ExBudgetState where
 newtype ExTally = ExTally (MonoidalHashMap ExBudgetCategory ExBudget)
     deriving stock (Eq, Generic, Show)
     deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
+    deriving anyclass NFData
 instance PrettyBy config ExTally where
     prettyBy config (ExTally m) =
         parens $ fold (["{ "] <> (intersperse (line <> "| ") $ fmap group $
           ifoldMap (\k v -> [(prettyBy config k <+> "causes" <+> prettyBy config v)]) m) <> ["}"])
 
-$(mtraverse makeLenses [''ExBudgetState, ''ExBudget])
-
 -- TODO See language-plutus-core/docs/Constant application.md for how to properly implement this
 estimateStaticStagedCost
     :: BuiltinName -> [WithMemory Value uni] -> (ExCPU, ExMemory)
 estimateStaticStagedCost _ _ = (1, 1)
 
+data CostingFunParameters =
+    CostingFunParameters
+    { paramAddInteger           :: CostingFunTwoArguments
+    , paramSubtractInteger      :: CostingFunTwoArguments
+    , paramMultiplyInteger      :: CostingFunTwoArguments
+    , paramDivideInteger        :: CostingFunTwoArguments
+    , paramQuotientInteger      :: CostingFunTwoArguments
+    , paramRemainderInteger     :: CostingFunTwoArguments
+    , paramModInteger           :: CostingFunTwoArguments
+    , paramLessThanInteger      :: CostingFunTwoArguments
+    , paramLessThanEqInteger    :: CostingFunTwoArguments
+    , paramGreaterThanInteger   :: CostingFunTwoArguments
+    , paramGreaterThanEqInteger :: CostingFunTwoArguments
+    , paramEqInteger            :: CostingFunTwoArguments
+    , paramConcatenate          :: CostingFunTwoArguments
+    , paramTakeByteString       :: CostingFunTwoArguments
+    , paramDropByteString       :: CostingFunTwoArguments
+    , paramSHA2                 :: CostingFunOneArgument
+    , paramSHA3                 :: CostingFunOneArgument
+    , paramVerifySignature      :: CostingFunThreeArguments
+    , paramEqByteString         :: CostingFunTwoArguments
+    , paramLtByteString         :: CostingFunTwoArguments
+    , paramGtByteString         :: CostingFunTwoArguments
+    , paramIfThenElse           :: CostingFunThreeArguments
+    }
+    deriving (Show, Eq, Generic, Lift)
+    deriving (FromJSON, ToJSON) via CustomJSON '[FieldLabelModifier (StripPrefix "param", CamelToSnake)] CostingFunParameters
```

Comment:
Can we have a comment about what determines the JSON format, please? Just looking pretty?

Reply:
And for the others. Maybe a separate Note then.

Reply:
Yeah, ease of creation :-)